go/vt/vtgate/resilient_srv_topo_server.go

```diff
@@ -299,47 +328,54 @@ func (server *ResilientSrvTopoServer) GetSrvKeyspace(ctx context.Context, cell,
 	entry.mutex.Lock()
 	defer entry.mutex.Unlock()
 
-	// If the entry exists, return it
-	if entry.value != nil {
-		return entry.value, nil
+	// If the watch is already running, return the value
+	if entry.watchRunning {
+		return entry.value, entry.lastError
 	}
 
-	// not in cache, get the real value
+	// watch is not running, let's try to start it
 	newCtx := context.Background()
-
-	// start watching
-	notifications, stopWatching, err := server.topoServer.WatchSrvKeyspace(newCtx, cell, keyspace)
+	notifications, err := server.topoServer.WatchSrvKeyspace(newCtx, cell, keyspace)
 	if err != nil {
+		// lastError and lastErrorCtx will be visible from the UI
+		// until the next try
+		entry.value = nil
 		entry.lastError = err
 		entry.lastErrorCtx = ctx
 		log.Errorf("WatchSrvKeyspace failed for %v/%v: %v", cell, keyspace, err)
-		return nil, entry.lastError
+		return nil, err
 	}
 	sk, ok := <-notifications
 	if !ok {
-		entry.lastError = fmt.Errorf("failed to receive from channel")
+		// lastError and lastErrorCtx will be visible from the UI
+		// until the next try
+		entry.value = nil
+		entry.lastError = fmt.Errorf("failed to receive initial value from topology watcher")
 		entry.lastErrorCtx = ctx
 		log.Errorf("WatchSrvKeyspace first result failed for %v/%v", cell, keyspace)
-		close(stopWatching)
```

Comment:
You have to replace this with a cancel on the context which you passed to WatchSrvKeyspace().

Otherwise, the watcher will continue to run and you leaked a Go routine.

Alternatively, you can keep it running and set "entry.watchRunning" already before you read from "notifications" in line 318/348.